src/backend/src/api/services/deployments_service.rs

```diff
@@ -27,17 +30,81 @@ impl DeploymentsService {
             .collect()
     }
 
-    pub async fn create_deployment(
+    pub async fn init_deployment(
         &mut self,
-        deployment: Deployment,
+        user_id: UserId,
+        sdl: String,
     ) -> Result<DeploymentId, ApiError> {
         let deployment_id = DeploymentId::new()
             .await
             .map_err(|e| ApiError::internal(&format!("Failed to create deployment id: {}", e)))?;
 
-        self.deployments_memory
-            .insert(deployment_id.clone(), deployment);
+        let deployment = Deployment::new(sdl, user_id, deployment_id.to_string());
+
+        self.deployments_memory.insert(deployment_id, deployment);
 
         Ok(deployment_id)
     }
+
+    pub fn update_deployment(
+        &mut self,
+        deployment_id: DeploymentId,
+        deployment_update: DeploymentUpdate,
+    ) -> Result<(), ApiError> {
+        self.update_deployment_state(deployment_id, deployment_update)
+    }
+
+    pub fn set_failed_deployment(
+        &mut self,
+        deployment_id: DeploymentId,
+        reason: String,
+    ) -> Result<(), ApiError> {
+        self.update_deployment_state(deployment_id, DeploymentUpdate::Failed(reason))
+    }
+
+    pub fn get_akash_deployment_info(
+        &self,
+        deployment_id: &DeploymentId,
+    ) -> Result<Option<u64>, ApiError> {
+        self.deployments_memory
+            .get(deployment_id)
+            .map(|deployment| deployment.get_akash_info())
+            .ok_or(ApiError::not_found(&format!(
+                "Deployment {} not found",
+                deployment_id
+            )))
+    }
+
+    pub fn close_deployment(&mut self, deployment_id: DeploymentId) -> Result<(), ApiError> {
+        self.update_deployment_state(deployment_id, DeploymentUpdate::Closed)
+    }
+
+    fn update_deployment_state(
+        &mut self,
+        deployment_id: DeploymentId,
+        deployment_update: DeploymentUpdate,
+    ) -> Result<(), ApiError> {
+        let mut deployment = self.deployments_memory.get(&deployment_id).ok_or_else(|| {
+            ApiError::not_found(&format!("Deployment {} not found", deployment_id))
+        })?;
+
+        if let DeploymentUpdate::Failed(_) = deployment.state() {
```

Comment:
`DeploymentUpdate::Failed(_)` must be `deployment_update`, otherwise it'll always fail. Same for `DeploymentUpdate::Closed` down a few lines.